docs/writing-benchmarks.md

```diff
@@ -0,0 +1,236 @@
+for details about the effect and restrictions of the annotation in Kotlin/JVM.
+
+In the provided example, the `@Setup` annotation is used to populate an `ArrayList` with integers from `0` up to a specified `size`.
+
+### @TearDown
+
+The `@TearDown` annotation is used to denote a method that's executed after the benchmarking method(s).
+This method is typically responsible for cleaning up or deallocating any resources or conditions that were initialized in the `@Setup` method.
+
+In Kotlin/JVM, you can specify when the teardown method should be executed, e.g, `@TearDown(Level.Iteration)`.
+Refer to [JMH documentation of Level](https://javadoc.io/doc/org.openjdk.jmh/jmh-core/latest/org/openjdk/jmh/annotations/Level.html)
+for details about available levels. In other targets, it operates always on `Trial` level, meaning the teardown method
+is executed once after the entire set of benchmark method iterations.
+
+The `@TearDown` annotation helps you avoid performance bias and ensures the proper maintenance of resources and the
+preparation of a clean environment for the next run. As with the `@Setup` method, the `@TearDown` method's
+execution time is not included in the final benchmark results.
+
+Refer to [JMH documentation of @TearDown](https://javadoc.io/doc/org.openjdk.jmh/jmh-core/latest/org/openjdk/jmh/annotations/TearDown.html)
+for details about the effect and restrictions of the annotation in Kotlin/JVM.
+
+In our example, the `cleanup` function annotated with `@TearDown` is used to clear our `ArrayList`.
+
+### @Benchmark
+
+The `@Benchmark` annotation is used to specify the methods that you want to measure the performance of.
+It's the actual test you're running. The code you want to benchmark goes inside this method.
+All other annotations are used to control different things in measuring operations of benchmark methods.
+
+Benchmark methods may include only a single [Blackhole](#blackhole) type as an argument, or have no arguments at all.
+It's important to note that in Kotlin/JVM benchmark methods must always be `public`.
+Refer to [JMH documentation of @Benchmark](https://javadoc.io/doc/org.openjdk.jmh/jmh-core/latest/org/openjdk/jmh/annotations/Benchmark.html)
+for details about restrictions for benchmark methods in Kotlin/JVM.
+
+In our example, the `benchmarkMethod` function is annotated with `@Benchmark`,
+which means the toolkit will measure the performance of the operation of summing all the integers in the list.
+
+### @BenchmarkMode
+
+The `@BenchmarkMode` annotation sets the mode of operation for the benchmark.
+
+Applying the `@BenchmarkMode` annotation requires specifying a mode from the `Mode` enum.
+In Kotlin/JVM, the `Mode` enum has several options, including `SingleShotTime`.
+
+Refer to [JMH documentation of Mode](https://javadoc.io/doc/org.openjdk.jmh/jmh-core/latest/org/openjdk/jmh/annotations/Mode.html)
+for details about available options. In other targets, only `Throughput` and `AverageTime` are available.
```

Comment:
As with `State`, I think it's worth describing the general case first (covering `Throughput` and `AverageTime` modes) and then mentioning that on JVM the set of available modes is broader.

Reply:
Done.